### PR TITLE
setImmediate

### PR DESCRIPTION
**Pull request by juj**
Add support for the Microsoft-proposed setImmediate() function for performant vsyncless updates when targeting IE/Edge, and if not supported, emulate/polyfill it via postMessage() (for debugging/testing purposes on other browsers).

### CODE REVIEW

File: site/source/docs/api_reference/emscripten.h.rst [hunk 1]
```diff
@@ -245,17 +245,18 @@ Functions
...
 	   - If ``mode`` is EM_TIMING_RAF, then updates are performed using the ``requestAnimationFrame`` function (with vsync enabled), and this value is interpreted as a "swap interval" rate for the main loop. The value of ``1`` specifies the runtime that it should render at every vsync (typically 60fps), whereas the value ``2`` means that the main loop callback should be called only every second vsync (30fps). As a general formula, the value ``n`` means that the main loop is updated at every n'th vsync, or at a rate of ``60/n`` for 60Hz displays, and ``120/n`` for 120Hz displays.
+	   - If ``mode`` is EM_TIMING_SETIMMEDIATE, then updates are performed using the ``setImmediate`` function, or if not available, emulated via ``postMessage``. See `setImmediate on MDN <https://developer.mozilla.org/en-US/docs/Web/API/Window/setImmediate>` for more information. Note that this mode is strongly not recommended to be used when deploying Emscripten output to the web, since it depends on an unstable web extension that is in draft status, and browsers other than IE do not currently support it.
```
Comment (kripken) [inline site/source/docs/api_reference/emscripten.h.rst]: please use all caps for STRONGLY NOT RECOMMENDED (or bold).

also please mention that it is controversial among other browsers (as the link mentions)

File: src/library_browser.js [hunk 2]
```diff
@@ -1072,15 +1072,36 @@ mergeInto(LibraryManager.library, {
...
       };
       Browser.mainLoop.method = 'timeout';
     } else if (mode == 1 /*EM_TIMING_RAF*/) {
-      Browser.mainLoop.scheduler = function Browser_mainLoop_scheduler() {
+      Browser.mainLoop.scheduler = function Browser_mainLoop_scheduler_rAF() {
         Browser.requestAnimationFrame(Browser.mainLoop.runner);
       };
       Browser.mainLoop.method = 'rAF';
+    } else if (mode == 2 /*EM_TIMING_SETIMMEDIATE*/) {
+      if (!window.setImmediate) {
```
Comment (kripken) [inline src/library_browser.js]: i think you need `['setImmediate']` as closure almost certainly isn't aware of that name

File: src/library_browser.js [hunk 3]
```diff
@@ -1072,15 +1072,36 @@ mergeInto(LibraryManager.library, {
...
+          }
+        }
+        window.addEventListener("message", Browser_setImmediate_messageHandler, true);
+        window.setImmediate = function Browser_emulated_setImmediate(func) {
+          setImmediates.push(func);
+          window.postMessage(emscriptenMainLoopMessageId, "*");
+        }
+      }
+      Browser.mainLoop.scheduler = function Browser_mainLoop_scheduler_setImmediate() {
+        window.setImmediate(Browser.mainLoop.runner);
```
Comment (kripken) [inline src/library_browser.js]: here too

**Comment by kripken** [comment]
lgtm with those.

File: site/source/docs/api_reference/emscripten.h.rst (hunk 1 above)
Comment (juj) [inline site/source/docs/api_reference/emscripten.h.rst]: Check.

File: src/library_browser.js (hunk 3 above)
Comment (juj) [inline src/library_browser.js]: Check.

**Comment by juj** [comment]
Merged, thanks for the review!

**Pull request merged**